addons/sourcemod/scripting/superzombiefortress.sp

```diff
@@ -979,55 +981,129 @@ public Action hook_JoinTeam(int client, const char[] command, int argc)
 
 	if (!zf_bEnabled) return Plugin_Continue;
 	if (argc < 1 && StrEqual(command, "jointeam", false)) return Plugin_Handled;
+	if (roundState() < RoundGrace) return Plugin_Continue;
 	
 	//Get command/arg on which team player joined
-	if (StrEqual(command, "spectate", false))
-		strcopy(cmd1, sizeof(cmd1), "spectate");
-	else
+	if (StrEqual(command, "jointeam", false)) // this is done because "jointeam spectate" should take priority over "spectate"
 		GetCmdArg(1, cmd1, sizeof(cmd1));
+	else if (StrEqual(command, "spectate", false))
+		strcopy(cmd1, sizeof(cmd1), "spectate");	
+	else if (StrEqual(command, "autoteam", false))
+		strcopy(cmd1, sizeof(cmd1), "autoteam");		
+	
+	//Check if client is trying to skip playing as zombie by joining spectator
+	if (StrEqual(cmd1, "spectate", false))
+		CheckZombieBypass(client);
+			
+	// Assign team-specific strings
+	if (zomTeam() == INT(TFTeam_Blue))
+	{
+		sSurTeam = "red";
+		sZomTeam = "blue";
+		sZomVgui = "class_blue";
+	}
+	else
+	{
+		sSurTeam = "blue";
+		sZomTeam = "red";
+		sZomVgui = "class_red";
+	}
 
-	if (roundState() >= RoundGrace)
+	if (IsClientInGame(client))
 	{
-		// Assign team-specific strings
-		if (zomTeam() == INT(TFTeam_Blue))
+		if (roundState() == RoundGrace)
 		{
-			sSurTeam = "red";
-			sZomTeam = "blue";
-			sZomVgui = "class_blue";
-		}
-		else
-		{
-			sSurTeam = "blue";
-			sZomTeam = "red";
-			sZomVgui = "class_red";
-		}
-		
-		//Check if client is trying to skip playing as zombie by joining spectator
-		if (StrEqual(cmd1, "spectate", false))
-			CheckZombieBypass(client);
+			int iTeam = GetClientTeam(client);
 		
-		// If client tries to join the survivor team or a random team
-		// during grace period or active round, place them on the zombie
-		// team and present them with the zombie class select screen.
-		if (StrEqual(cmd1, sSurTeam, false) || StrEqual(cmd1, "auto", false))
-		{
-			ChangeClientTeam(client, zomTeam());
-			ShowVGUIPanel(client, sZomVgui);
-			return Plugin_Handled;
-		}
-		// If client tries to join the zombie team or spectator
-		// during grace period or active round, let them do so.
-		else if (StrEqual(cmd1, sZomTeam, false) || StrEqual(cmd1, "spectate", false))
-		{
-			return Plugin_Continue;
+			// If a client tries to join the infected team or a random team during grace period...
+			if (StrEqual(cmd1, sZomTeam, false) || StrEqual(cmd1, "auto", false) || StrEqual(cmd1, "autoteam", false))
+			{
+				// ...as survivor, don't let them.
+				if (iTeam == 2)
+				{
+					CPrintToChat(client, "{red}You can not switch to the opposing team during grace period.");
+					return Plugin_Handled;
+				}
+
+				// ...as a spectator who didn't start as an infected, set them as infected after grace period ends, after warning them.
+				if (iTeam <= 1 && !g_bStartedAsZombie[client])
+				{
+					if (!g_bWaitingForTeamSwitch[client])
+					{
+						if (iTeam == 0) // If they're unassigned, let them spectate for now.
```

Comment:
Is it even possible to reach this check? Or if that makes any changes. Even then should use `TFTeam` enum instead of 0.
I'm fine with having `iTeam <= 1` as both spectator team check btw.

Reply:
it is possible, think of someone who just joined the server just trying to join a team, they'll be unassigned during that time, rather than an spectator
What it changes is that unassigned can only cycle through map observer points, and spectators can freecam/cycle through other players. This is just a minor thing for player convenience